apps/src/code-studio/components/header/EditableProjectName.tsx

```diff
@@ -0,0 +1,69 @@
+import React, {useEffect, useState} from 'react';
+import {useDispatch, useSelector} from 'react-redux';
+import {refreshProjectName, ProjectState} from '../../projectRedux';
+import EditProjectName from './EditProjectName';
+import DisplayProjectName from './DisplayProjectName';
+import {LabState} from '@cdo/apps/lab2/lab2Redux';
+import Lab2Registry from '@cdo/apps/lab2/Lab2Registry';
+
+const EditableProjectName: React.FunctionComponent<
+  EditableProjectNameProps
+> = ({onChangedWidth}) => {
+  const [editName, setEditName] = useState(false);
+  const legacyProjectName = useSelector(
+    (state: {project: ProjectState}) => state.project.projectName
+  );
+  const lab2ProjectName = useSelector(
+    (state: {lab: LabState}) => state.lab.channel?.name
+  );
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (onChangedWidth) {
+      onChangedWidth();
+    }
+  });
+
+  function beginEdit() {
+    setEditName(true);
+  }
+
+  function finishEdit() {
+    setEditName(false);
+    if (!Lab2Registry.hasEnabledProjects()) {
+      dispatch(refreshProjectName());
+    }
+  }
+
+  function saveProjectName(newName: string) {
+    if (Lab2Registry.hasEnabledProjects()) {
```

Comment:
Since we're checking `hasEnabledProjects` a couple times here, would it be worth it to abstract this out one layer further, and move the check up there? Maybe we have a `Lab2EditableProjectName` and a `LegacyEditableProjectName`, each which internally doesn't need to make any checks, then a wrapper that's basically just `if (hasEnabledProjects) { return <Lab2EditableProjectName/> } else { return <LegacyEditableProjectName/>`. Could be overkill, but just to help minimize the amount of lab2-specific conditional logic in these components

Reply:
There's an advantage to doing it this way, which is we don't do the check until the user clicked the button, so we are going to be done loading and the project manager will exist for Lab2. I had tried something similar for a previous refactor and ran into issues where due to load times it was falling back to legacy because at the time we did the check we didn't have a project manager yet. I do see the advantage of simplifying this, but it's only a couple checks in this file.

Reply:
Ah okay gotcha, makes sense!